// reconlx/src/pagination/pagination.ts
import {
    ButtonInteraction,
    MessageActionRow,
    MessageButton,
    MessageComponentType,
    MessageEmbed,
} from "discord.js";
import { ButtonNames, PaginationOptions } from "./pagination.interfaces";

const defaultEmojis = {
    first: "⬅️",
    previous: "◀️",
    next: "▶️",
    last: "➡️",
};

export const pagination = async (options: PaginationOptions) => {
    const { message, embeds, button, time, max, customFilter } = options;
    let currentPage = 1;

    const getButtonData = (name: ButtonNames) => {
        return button.find((btn) => btn.name === name);
    };

    const generateButtons = (state?: boolean) => {
        const checkState = (name: ButtonNames) => {
            if (
                (["first", "previous"] as ButtonNames[]).includes(name) &&
                currentPage === 1
            )
                return true;

            if (
                (["next", "last"] as ButtonNames[]).includes(name) &&
                currentPage === embeds.length
            )
                return true;

            return false;
        };

        const names: ButtonNames[] = ["first", "previous", "next", "last"];
        return names.reduce(
            (accumulator: MessageButton[], name: ButtonNames) => {
                accumulator.push(
                    new MessageButton()
                        .setEmoji(
                            getButtonData(name)?.emoji || defaultEmojis[name]
                        )
                        .setCustomId(name)
                        .setDisabled(state || checkState(name))
                );
                return accumulator;
            },
            []
        );
    };

    const components = (state?: boolean) => [
        new MessageActionRow().addComponents(generateButtons(state)),
    ];

    const changeFooter = (embed: MessageEmbed) => {
        if (embed.footer.text) {
            return embed.setFooter(
                `${embed.footer.text} - Page ${currentPage} of ${embeds.length}`,
                embed.footer.iconURL
            );
        }
        return embed.setFooter(`Page ${currentPage} of ${embeds.length}`);
    };

    await message.channel.send({
        embeds: [changeFooter(embeds[0])],
        components: components(),
    });

    const defaultFilter = (interaction: ButtonInteraction) =>
        interaction.user.id === message.author.id;

    const filter = customFilter || defaultFilter;

    const collectorOptions = () => {
        const opt = {
            filter,
            componentType: "BUTTON" as MessageComponentType,
        };

        if (max) opt["max"] = max;
        if (time) opt["time"] = time;

        return opt;
    };

    const collector = message.channel.createMessageComponentCollector(
        collectorOptions()
    );

    collector.on("collect", (interaction) => {
        const id = interaction.customId as ButtonNames;

        if (id === "first") currentPage = 1;
        if (id === "previous") currentPage--;
        if (id === "next") currentPage++;
        if (id === "last") currentPage === embeds.length;

        interaction.update({
            embeds: [changeFooter(embeds[currentPage])],
            components: components(),
        });
    });

    collector.on("end", (interactions) => {
        interactions.first().update({
            components: components(true),
        });
    });
};
